app/lib/actions/trigger_proxy_batch.rb

```diff
@@ -0,0 +1,77 @@
+module Actions
+  # This action plans proxy tasks in batches.
+  # It needs to be manually notified about the next batch being available by sending a TriggerNextBatch event.
+  #
+  # The ProxyAction needs to be planned with `:use_batch_triggering => true` to activate the feature
+  class TriggerProxyBatch < Base
+    TriggerNextBatch = Struct.new(:batches)
```

Comment:
Hm, that might have been a bad suggestion. Looks like the serializer in dynflow is not able to properly serialize a struct.

```
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: 2021-08-24T12:59:57.824Z 4357 TID-51vx WARN: {"context":"Job raised exception","job":{"queue":"remote_execution","args":[{"class":"Dynflow::Director::StepWorkItem","execution_plan_id":"a331b357-a03c-4090-893c-6925b3a65bdd","queue":"remote_execution","sender_orchestrator_id":"b2073151-6f02-45cc-ab5c-2c59cc967ad8","step":{"execution_plan_uuid":"a331b357-a03c-4090-893c-6925b3a65bdd","id":2,"state":"pending","class":"Dynflow::ExecutionPlan::Steps::RunStep","action_class":"Actions::RemoteExecution::RunHostsJob","action_id":1,"error":null,"started_at":null,"ended_at":null,"execution_time":0.0,"real_time":0.0,"progress_done":0,"progress_weight":1.0,"queue":"remote_execution"}}],"class":"Dynflow::Executors::Sidekiq::WorkerJobs::PerformWork","retry":false,"backtrace":true,"jid":"7a5f478cf29fe634a1dcbb0e","created_at":1629809965.3489962,"enqueued_at":1629809965.3490453},"jobstr":"{\"queue\":\"remote_execution\",\"args\":[{\"class\":\"Dynflow::Director::StepWorkItem\",\"execution_plan_id\":\"a331b357-a03c-4090-893c-6925b3a65bdd\",\"queue\":\"remote_execution\",\"sender_orchestrator_id\":\"b2073151-6f02-45cc-ab5c-2c59cc967ad8\",\"step\":{\"execution_plan_uuid\":\"a331b357-a03c-4090-893c-6925b3a65bdd\",\"id\":2,\"state\":\"pending\",\"class\":\"Dynflow::ExecutionPlan::Steps::RunStep\",\"action_class\":\"Actions::RemoteExecution::RunHostsJob\",\"action_id\":1,\"error\":null,\"started_at\":null,\"ended_at\":null,\"execution_time\":0.0,\"real_time\":0.0,\"progress_done\":0,\"progress_weight\":1.0,\"queue\":\"remote_execution\"}}],\"class\":\"Dynflow::Executors::Sidekiq::WorkerJobs::PerformWork\",\"retry\":false,\"backtrace\":true,\"jid\":\"7a5f478cf29fe634a1dcbb0e\",\"created_at\":1629809965.3489962,\"enqueued_at\":1629809965.3490453}"}
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: 2021-08-24T12:59:57.824Z 4357 TID-51vx WARN: RuntimeError: Missing class key in {:batches=>5}
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: 2021-08-24T12:59:57.824Z 4357 TID-51vx WARN: /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.9/lib/dynflow/serializer.rb:66:in `generate_other'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/algebrick-0.7.3/lib/algebrick/serializer.rb:41:in `dump'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.9/lib/dynflow/serializer.rb:28:in `dump'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/algebrick-0.7.3/lib/algebrick/serializer.rb:122:in `block in generate_value'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/algebrick-0.7.3/lib/algebrick/serializer.rb:121:in `each'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/algebrick-0.7.3/lib/algebrick/serializer.rb:121:in `inject'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/algebrick-0.7.3/lib/algebrick/serializer.rb:121:in `generate_value'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/algebrick-0.7.3/lib/algebrick/serializer.rb:37:in `dump'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.9/lib/dynflow/serializer.rb:28:in `dump'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.9/lib/dynflow/serializer.rb:26:in `block in dump'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.9/lib/dynflow/serializer.rb:26:in `collect'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.9/lib/dynflow/serializer.rb:26:in `dump'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.9/lib/dynflow/executors/sidekiq/serialization.rb:8:in `serialize'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.9/lib/dynflow/executors/sidekiq/serialization.rb:20:in `block in client_push'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.9/lib/dynflow/executors/sidekiq/serialization.rb:20:in `map'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.9/lib/dynflow/executors/sidekiq/serialization.rb:20:in `client_push'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/worker.rb:93:in `perform_async'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.9/lib/dynflow/executors/sidekiq/worker_jobs.rb:18:in `perform'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.9/lib/dynflow/executors/sidekiq/serialization.rb:27:in `perform'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:192:in `execute_job'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:165:in `block (2 levels) in process'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/middleware/chain.rb:128:in `block in invoke'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/middleware/chain.rb:133:in `invoke'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:164:in `block in process'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:137:in `block (6 levels) in dispatch'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/job_retry.rb:109:in `local'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:136:in `block (5 levels) in dispatch'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq.rb:37:in `block in <module:Sidekiq>'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:132:in `block (4 levels) in dispatch'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:250:in `stats'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:127:in `block (3 levels) in dispatch'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/job_logger.rb:8:in `call'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:126:in `block (2 levels) in dispatch'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/job_retry.rb:74:in `global'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:125:in `block in dispatch'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/logging.rb:48:in `with_context'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/logging.rb:42:in `with_job_hash_context'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:124:in `dispatch'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:163:in `process'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:83:in `process_one'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:71:in `run'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/util.rb:16:in `watchdog'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/sidekiq-5.2.7/lib/sidekiq/util.rb:25:in `block in safe_thread'
Aug 24 08:59:57 satellite-610s14.example.com dynflow-sidekiq@worker-1[4357]: /opt/theforeman/tfm/root/usr/share/gems/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
```

Reply:
Back to algebrick then? I'd have to go with Hash, or implement the serializer. I'd rather not....

Reply:
Iirc a "proper" class should work. Neither that nor algebrick is pretty, so feel free to pick one